### PR TITLE
[refactor] #3349: Remove `EvaluateOnHost` from smart contracts

### LINKED ISSUE #3349: Host evaluation is largely redundant for instructions
We should strive to minimize interactions of wasm smart contract and host environment. The reason why a `wasm` smart contract has to call a host defined function to evaluate an expression is because it needs access to the `wsv`. However, of all the implementations of trait `Evaluate` in `core/smartcontracts`, `wsv` is only required for executing an `Expression::Query`.

`EvaluateOnHost` should leverage this by not doing a host call unless absolutely required

### PR DESCRIPTION
**Pull request by mversic**
## Description

Evaluation of expressions on the host is unnecessary so it's been moved into the smart contract itself. `wsv` is only used in one expression evaluation and that is `Expression::Query`. Had we not had this expression, every expression could be evaluated deterministically(whatever the current state of the world may be)

This should not bloat the binary once dynamic linking is in place because evaluation will be taking place in the dynamically linked wasm not the smart contract itself

### Linked issue

Closes #3349

### Benefits

* the less interaction with the host the better the performance

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: data_model/src/evaluate.rs [hunk 1]
```diff
@@ -1,105 +1,107 @@
...
-use super::Evaluate;
-use crate::{prelude::ValidQuery, wsv::WorldStateView};
+/// Context of expression evaluation, holding (name, value) pairs for resolving identifiers
+pub trait Context: Clone {
+    /// Execute query against the current state of `Iroha`
+    ///
+    /// # Errors
+    ///
+    /// If query evaluation/execution fails
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure>;
```
Comment (Arjentix) [inline data_model/src/evaluate.rs:22]: Okay, I see, you did it so that we can execute query from WASM without direct wsv access... But it still doesn't make sense to me from logic side of view

File: wasm/src/lib.rs [hunk 2]
```diff
@@ -85,40 +87,33 @@ impl ExecuteOnHost for data_model::query::QueryBox {
...
 
-    fn evaluate(&self) -> Result<Self::Value, Self::Error> {
-        #[cfg(not(test))]
-        use host::evaluate_on_host as host_evaluate_on_host;
-        #[cfg(test)]
-        use tests::_iroha_wasm_evaluate_on_host_mock as host_evaluate_on_host;
+impl iroha_data_model::evaluate::Context for Context {
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure> {
+        Ok(query.execute())
+    }
```
Comment (Arjentix) [inline wasm/src/lib.rs:109]: I can't see a reason to not to allow user to just call `query.execute()` in its own. So why do we need this?

File: data_model/src/evaluate.rs [hunk 3]
```diff
@@ -1,105 +1,107 @@
...
-            Contains(contains) => contains.evaluate(wsv, context)?,
-            ContainsAll(contains_all) => contains_all.evaluate(wsv, context)?,
-            ContainsAny(contains_any) => contains_any.evaluate(wsv, context)?,
-            Where(where_expression) => where_expression.evaluate(wsv, context)?,
-            ContextValue(context_value) => context_value.evaluate(wsv, context)?,
-            Multiply(multiply) => multiply.evaluate(wsv, context)?,
-            Divide(divide) => divide.evaluate(wsv, context)?,
-            Mod(modulus) => modulus.evaluate(wsv, context)?,
-            RaiseTo(raise_to) => raise_to.evaluate(wsv, context)?,
+            Query(query) => context.query(query)?,
```
Comment (mversic) [inline data_model/src/evaluate.rs:72]: because you cannot execute query from `data_model`

File: wasm/src/lib.rs (hunk 2 above)
Comment (mversic) [reply wasm/src/lib.rs:109]: look [here](https://github.com/hyperledger/iroha/pull/3351/files#r1157745952)

File: core/src/smartcontracts/mod.rs [hunk 4]
```diff
@@ -48,16 +35,44 @@ pub trait ValidQuery: Query {
...
+            values: BTreeMap::new(),
+            wsv,
+        }
+    }
+}
+
+impl iroha_data_model::evaluate::Context for Context<'_> {
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure> {
+        query.execute(self.wsv)
+    }
```
[the remark this answers is not in the document]
Comment (mversic) [reply core/src/smartcontracts/mod.rs:69]: I don't think it can be any other way. If this is not ok for you would you like to:

1. rename `Context` to something else?
2 entirely remove `Expression::Query`?

File: data_model/src/evaluate.rs [hunk 5]
```diff
@@ -1,105 +1,107 @@
...
+    ///
+    /// # Errors
+    ///
+    /// If query evaluation/execution fails
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure>;
+
+    /// Return a reference to the [`Value`] corresponding to the [`Name`].
+    fn get(&self, name: &Name) -> Option<&Value>;
+
+    /// Gets an iterator over the values in this context, in order by [`Name`].
```
Comment (QuentinI) [inline data_model/src/evaluate.rs:27]: This seems incorrect

**Review by QuentinI — CHANGES_REQUESTED** [review]

File: core/src/smartcontracts/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (QuentinI) [reply core/src/smartcontracts/mod.rs:69]: > Context can execute Query? This doesn't make sense to me

I would rephrase it as "context can be queried", which sounds reasonable

File: data_model/src/evaluate.rs [hunk 6]
```diff
@@ -1,105 +1,107 @@
...
     expression::{prelude::*, Expression},
     isi::error::{InstructionExecutionFailure as Error, MathError},
     prelude::*,
-    query::error::FindError,
+    query::error::{FindError, QueryExecutionFailure},
 };
 
-use super::Evaluate;
-use crate::{prelude::ValidQuery, wsv::WorldStateView};
+/// Context of expression evaluation, holding (name, value) pairs for resolving identifiers
```
Comment (QuentinI) [inline data_model/src/evaluate.rs:15]: Could you maybe expand on this? It's not obvious neither what are the identifiers we're talking about nor why we need them

File: core/src/smartcontracts/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mversic) [reply core/src/smartcontracts/mod.rs:69]: The difference is that Context now includes state of Iroha at the time of executing a query, not only identifiers found in the where clause of the expression.

File: data_model/src/evaluate.rs (hunk 6 above)
Comment (mversic) [reply data_model/src/evaluate.rs]: you are correct, it can be a pain to figure out what `Context` does. I've updated the docs now with a neat example

File: core/src/smartcontracts/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mversic) [reply core/src/smartcontracts/mod.rs:69]: We could evaluate all expressions on the client side had it not been for the `Expression:Query`. Atm, it is the only expressions that needs `wsv`

File: wasm/src/lib.rs (hunk 2 above)
Comment (Arjentix) [reply wasm/src/lib.rs:109]: Okay, makes sense

File: core/src/smartcontracts/isi/mod.rs [hunk 7]
```diff
@@ -267,7 +266,7 @@ impl Execute for SetKeyValueBox {
             IdBox::DomainId(id) => {
                 SetKeyValue::<Domain, Name, Value>::new(id, key, value).execute(authority, wsv)
             }
-            _ => Err(Error::Unsupported(InstructionType::SetKeyValue)),
+            _ => Err(Error::Evaluate(InstructionType::SetKeyValue.into())),
```
Comment (Arjentix) [inline core/src/smartcontracts/isi/mod.rs:269]: Why so? I think `Unsupported` error clearly shows that this instructions with this parameter types is not supported. I don't quite understand what `Evaluate` error means here. Maybe it least something like `Evaluate(Unsupported(...))`?

File: data_model/src/isi.rs [hunk 8]
```diff
@@ -757,50 +757,63 @@ pub mod error {
...
-            /// Failed to validate.
-            #[display(fmt = "Failed to validate: {_0}")]
-            Validate(#[cfg_attr(feature = "std", source)] ValidationError),
+        }
+
+        /// Type error
+        #[derive(Debug, Display, Clone, PartialEq, Eq, FromVariant)]
+        // TODO: Only temporarily opaque because of problems with FFI
+        #[ffi_type(opaque)]
+        pub enum EvaluationError {
```
Comment (Arjentix) [inline data_model/src/isi.rs:807]: Okay, I see a comment about type here. Maybe rename it somehow? I.e. `TypeEvaluationError`?

File: data_model/src/evaluate.rs [hunk 9]
```diff
@@ -1,299 +1,277 @@
...
+    fn get(&self, name: &Name) -> Option<&Value>;
+
+    /// Update this context with given values.
+    fn update(&mut self, other: impl IntoIterator<Item = (Name, Value)>);
+}
+
+/// Calculate the result of the expression without mutating the state.
+pub trait Evaluate {
+    /// The resulting type of the expression.
+    type Value;
```
Comment (Arjentix) [inline data_model/src/evaluate.rs:51]: Isn't it always `iroha_data_model::Value`?

File: permission_validators/account/remove_key_value/src/lib.rs [hunk 10]
```diff
@@ -44,7 +44,7 @@ pub fn validate(authority: <Account as Identifiable>::Id, instruction: Instructi
     };
 
     let IdBox::AccountId(account_id) = remove_key_value.object_id()
-        .evaluate()
+        .evaluate(&Context::new())
```
Comment (Arjentix) [inline permission_validators/account/remove_key_value/src/lib.rs:47]: What about:

```rust
pub trait Evaluate {
    // ...
    
    fn evaluate_default<C: Context + Default>(&self) -> Result<Self::Value, Error> {
        Evaluate::evaluate(self, Context::default())
    }
}
```

**Review by Arjentix — DISMISSED** [review]

File: data_model/src/evaluate.rs (hunk 9 above)
Comment (mversic) [reply data_model/src/evaluate.rs:51]: no, `EvaluatesTo<V>` evaluates to a concrete type `V`

File: core/src/smartcontracts/isi/mod.rs (hunk 7 above)
Comment (mversic) [reply core/src/smartcontracts/isi/mod.rs:269]: IMO `Unsupported` is a type of evaluation error because your expression failed to produce a valid type. `Evaluate` is about whether the instruction is valid in the sense that there is at least one state of Iroha that it can be executed against. If evaluation fails this means that instruction is not a valid construct according to the specification of Iroha DSL. 

`Validate` on the other hand is a separate step that verifies that a valid instruction can be executed at this time on the current state of Iroha. This is about permissions or invalid operations like divide amount in metadata by 0, etc.

Interestingly, `Expression::Query` is the only expression that requires `wsv`. If we wanted to pre-evaluate instructions we could do so by reducing every expression either to a value or to a `f(query)`. This would bring us a benefit of smaller chain size and less repeated calculation. In this case we can even consider client side evaluation

File: data_model/src/isi.rs (hunk 8 above)
Comment (mversic) [reply data_model/src/isi.rs:807]: I'm closing this comment in favor of the one before

File: permission_validators/account/remove_key_value/src/lib.rs (hunk 10 above)
Comment (mversic) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: that is a good idea

Comment (mversic) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: it's not possible to implement `Context::default` because it needs to get a reference to `wsv`

**Review by QuentinI — APPROVED** [review]

File: permission_validators/account/remove_key_value/src/lib.rs (hunk 10 above)
Comment (Arjentix) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: Yes, but possible for wasm:: Context, I think. Because it doesn't need wsv. And that would be enough to make use out of this function

File: config/src/wasm.rs [hunk 11]
```diff
@@ -3,7 +3,7 @@
 use iroha_config_base::derive::{Documented, Proxy};
 use serde::{Deserialize, Serialize};
 
-const DEFAULT_FUEL_LIMIT: u64 = 20_000_000;
+const DEFAULT_FUEL_LIMIT: u64 = 25_000_000;
```
Comment (Arjentix) [inline config/src/wasm.rs:6]: So we do more computations on the WASM side without extra communication with the host. That's why it increased?

File: permission_validators/account/remove_key_value/src/lib.rs (hunk 10 above)
Comment (mversic) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: mind you, it's part of the trait. I can't implement `Default` for `core::Context`

File: config/src/wasm.rs (hunk 11 above)
Comment (mversic) [reply config/src/wasm.rs:6]: yes, it should reduce with my next PR where I would reduce computation on wasm side

File: permission_validators/account/remove_key_value/src/lib.rs (hunk 10 above)
Comment (Arjentix) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: You don't have to make `Default` to be a part of `Context` trait. You can just add this function with additional `Default` bound to the `Evaluate` trait

Comment (Arjentix) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: It will be impossible to call this function for `core::Context` but will be possible for `wasm:: Context`

Comment (mversic) [reply permission_validators/account/remove_key_value/src/lib.rs:47]: let's leave it for my next PR, it'll make a difference there
